### PR TITLE
Patch for Issue #102

### PR DESCRIPTION
**Pull request by okeeble**
Patch supplied by Andrey Kiryanov for issue #102, "Gridftp failures between ipv4 client and dual-stack server".

This has been installed on a couple of DPM production sites for over a month.

### CODE REVIEW

**Comment by fscheiner** [comment]
I created a patch that also updates the `globus-gridftp-server(8)` manpage and the packaging information accordingly.

@ellert 
I bumped the minor version only, but maybe a bump of the major and minor version would be more appropriate, as functionality has changed (incl. globus-gridftp-server option)? Should we also change the minimum version numbers for the requirements?

****

Patch is attached. BTW, is there an easier way, e.g. like adding that patch as commit to Oliver's branch, just like I could rebase his branch earlier?

[updates-to-manpage-and-packaging-metadata.patch.txt](https://github.com/gridcf/gct/files/3482585/updates-to-manpage-and-packaging-metadata.patch.txt)

**Comment by fscheiner** [comment]
I'd rephrase the commit message to something like:

```
Fix problems between dual-stack (IPv4/IPv6) servers and IPv4-only clients

Fixes #102 
```

After merging this should still automatically close #102, but the "important" information is in the first line.

**Comment by fscheiner** [comment]
Just spotted [this comment in the CERN JIRA](https://its.cern.ch/jira/browse/LCGDM-2817?focusedCommentId=2681195&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2681195):

```
[...]
Note that one side-effect of this patch is that an ipv4 client will not be able
to establish an ipv6 transfer between a patched DPM and another ipv6
capable endpoint.
```

@ellert @matyasselmeci @msalle @okeeble 
Did the above described operation work in the past - i.e. control channels established via IPv4 and data channels established via IPv6 IIUIC? Does it constitute a problem when it no longer works with the patch included?

...ok, it's not enabled by default, so maybe not an issue.

**Comment by matyasselmeci** [comment]
Does this change conflict with the epsv_ip option ("Adds an IPv6 address to EPSV response. Breaks RFC 2428, but allows redirection to work with IPv6.")? We have that option set by default for OSG sites that use gridftp (and I was hoping to make that the default in GCT if it doesn't cause any problems).

**Comment by okeeble** [comment]
Thanks for taking a look at this.
@fscheiner - I don't know if this scenario ever happened in reality. With this patch the transfer would be established over ipv4, so things should not break.
@matyasselmeci - no, there's no conflict, but one could say this patch extends the scope of epsv_ip. It's the return of an address that breaks RFC 2428, and now we're allowing ourselves to return an ipv4 address under certain circumstances.

**Comment by ellert** [comment]
> I bumped the minor version only, but maybe a bump of the major and minor version would be more appropriate, as functionality has changed (incl. globus-gridftp-server option)? Should we also change the minimum version numbers for the requirements?

If the change adds new functions to the API, the major version AND age should both be increased by one (which keeps the soname unchanged), and the minor version reset to 0. If there are dependent packages that use the additions to the API their (build)requirements should be increased accordingly (in configure.ac, *.spec and debian/control). The package where the major version changes should also change its debian/*.shlibs.

globus-gridftp-server-control must increase the major version. For globus-gridftp-server it is not necessary.

> Patch is attached. BTW, is there an easier way, e.g. like adding that patch as commit to Oliver's branch, just like I could rebase his branch earlier?

If you read carefully it says "Add more commits by pushing to the master branch on okeeble/gct."

**Comment by okeeble** [comment]
@ellert - what we've done is add a runtime configuration option, which required adding one new function to globus_gridftp_server_control. Its API has therefore indeed changed, but in reality this is internal and noone needs to change any code. I just want to check that this is understood before we bump a major version.

**Comment by ellert** [comment]
Bumping the major version together with the age means that the new library keeps the same soname and is backwards compatible. Anything that was linked to the previous version continues to work with the new version without recompiling or relinking. What the change in major version indicates is that everything linked to the new version might not work with the old version, which is true if it uses the new function.

We are not bumping the major version and resetting age to 0 here, which would be the right thing to do if we REMOVED something from the API or changed it in some backward incompatible way. In that case the soname changes and everything needs to be recompiled/relinked.

**Comment by okeeble** [comment]
Thanks - makes perfect sense.

**Comment by fscheiner** [comment]
@ellert 
Sorry for the delay, I was busy with other things last week.

> > I bumped the minor version only, but maybe a bump of the major and minor version would be more appropriate, as functionality has changed (incl. globus-gridftp-server option)? Should we also change the minimum version numbers for the requirements?
> 
> If the change adds new functions to the API, the major version AND age should both be increased by one (which keeps the soname unchanged), and the minor version reset to 0. If there are dependent packages that use the additions to the API their (build)requirements should be increased accordingly (in configure.ac, *.spec and debian/control). The package where the major version changes should also change its debian/*.shlibs.
> 
> globus-gridftp-server-control must increase the major version. For globus-gridftp-server it is not necessary.

Ok, thanks for the explanation. I hope I did it correctly now.

> > Patch is attached. BTW, is there an easier way, e.g. like adding that patch as commit to Oliver's branch, just like I could rebase his branch earlier?
> 
> If you read carefully it says "Add more commits by pushing to the master branch on okeeble/gct."

I see. Before I try to push my modified patch, can you please have a look if my library version changes and changes to the packaging metadata are correct now?

[updates-to-manpage-and-packaging-metadata-v2.patch.txt](https://github.com/gridcf/gct/files/3516617/updates-to-manpage-and-packaging-metadata-v2.patch.txt)

If they're correct, I'll then try to push directly to the master branch of Oliver's fork of the GCT.

**Comment by ellert** [comment]
> I see. Before I try to push my modified patch, can you please have a look if my library version changes and changes to the packaging metadata are correct now?
> 
> [updates-to-manpage-and-packaging-metadata-v2.patch.txt](https://github.com/gridcf/gct/files/3516617/updates-to-manpage-and-packaging-metadata-v2.patch.txt)
> 
> If they're correct, I'll then try to push directly to the master branch of Oliver's fork of the GCT.

This looks OK.
If I run `globus-gridftp-server -help -asciidoc > globus-gridftp-server-options.txt`, the resulting file has 4 additional spaces on the line after "\*-epsv-match\*::" that are not present if I apply your patch. It is maybe not important - the manpage doesn't change due to those spaces.

**Comment by fscheiner** [comment]
@ellert 
> > If they're correct, I'll then try to push directly to the master branch of Oliver's fork of the GCT.
> 
> This looks OK.
> If I run `globus-gridftp-server -help -asciidoc > globus-gridftp-server-options.txt`, the resulting file has 4 additional spaces on the line after "*-epsv-match*::" that are not present if I apply your patch. It is maybe not important - the manpage doesn't change due to those spaces.

Ok, I see, thanks for checking, I now use the output of `globus-gridftp-server [...]` instead of editing the file directly (my editor didn't highlight white space so I didn't notice).

****

Will merge the PR when Travis is finished.
